src/development/cplusplus-codeguide.rst

```diff
@@ -0,0 +1,296 @@
+.. doctest-skip-all
+.. _code-guide:
+
+.. todo::
+    - Pretty much everythin
+    - Actual code guides
+
+************************
+C++ Coding Guidelines
+************************
+This section describes requirements and guidelines for development and testing of a new C++ project on GitLab. The COntinuous Integration tools are GitLab specific - but most of the processes could be easily moved to a Jenkins Pipeline if required. 
+
+.. contents:: Table of Contents
+
+An Example C++ Project
+======================
+
+We have created a `skeleton C++ project
+<https://gitlab.com/steve-ord/cpp_template>`_. Which should provide a full introduction to the various
+recommendations and requirements for the development of C++. The philosophy behind the development of this
+template was to demonstrate one way to meet the project guidelines. There are probably as many ways to organise C++ projects
+as there are developers there are sure to be some contraversial design descisions made. 
+
+This projects demonstrates a recommended
+project layout. It also demonstrates how to implement the following recommended C++ project development features. 
+
+* Continous integration (CI) setup using Gitlab 
+* Cmake as a build tool.
+* GoogleTest framework and example unit testing.
+* C++ linting using clang for stylistic errors.
+* Also test running under valgrind for memory errors.
+* gcov to measure test coverage 
+
+All building and testing is done within a docker container.
+
+Project Layout
+--------------
+Top Level
+^^^^^^^^^
+We have a top level of the project containing:
+
+CMakeLists.txt
+""""""""""""""
+        We have built this template using CMake. The structure of this file
+        will be discussed in `Building the Project`_.   
+
+LICENSE
+"""""""
+        All projects *must* have a license. We have include the recommended BSD
+        3 clause license template *please fill it in*.
```

Comment:
Do we require the licecne on every file, just headers, or is this one LICENCE file enough?

Reply:
I think the policy is one licence @flyingfrog81 ?

Reply:
The overall requirement at project level is to have one LICENSE file and possibly one COPYRIGHT file when having multiple Copyright holders.
Do you feel like we should extend it for C++ in having explicit headers stating LICENSE and copyright? I always find it to cause misalignments unless a good degree of automation is implemented to realise it.

Reply:
I don't like the license/copyright per file. I have it for one project and it just not get maintained. We decided to move to a single license for our most recent projects. Of course it is possible for people to pull individual files - but if they are the sort of person to do that  and not acknowledge then they would have stripped the top off your file anyway.